### DISCUSSION TITLE
Cannot run Script: Job problem

### DISCUSSION THREAD

**Discussion by drobnymichal** [netbox-community/netbox#13041] [Help Wanted!]
I am trying to run a script via the Web interface. The script is PENDING.

The problem is probably in the netbox-worker.
This is the log message I got: (netbox-worker)
```
15:46:04 [Job 019abc10-ccd7-41d8-8d7e-2e0b64cb3a5a]: exception raised while executing (extras.scripts.run_script)
Traceback (most recent call last):
    File "/opt/netbox/venv/lib/python3.10/site-packages/rq/worker.py", line 1428, in perform_job                                                              
        rv = job.perform()
    File "/opt/netbox/venv/lib/python3.10/site-packages/rq/job.py", line 1278, in perform
        self._result = self._execute()                                            
    File "/opt/netbox/venv/lib/python3.10/site-packages/rq/job.py", line 1315, in _execute
        result = self.func(*self.args, **self.kwargs
    File "/opt/netbox/netbox/extras/scripts.py", line 478, in run_script
        script = module.scripts.get(job.name)()
TypeError: 'NoneType' object is not callable 
```

CLI run is successful.

Do you have any idea how to fix it?

Thank you for reply.

**Comment by candlerb** [comment]
There is a suggestion for how to run a script standalone (outside of the web interface) here: #12766

In short, add the following to the top of your script:

```
#!/opt/netbox/venv/bin/python
import django, os, sys
sys.path.append('/opt/netbox/netbox')
os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'netbox.settings')
django.setup()
```

Then run it like this:

```
/opt/netbox/venv/bin/python yourscript.py
```

This should give you a better error message than Netbox itself gives.

**Reply by drobnymichal** [reply]
Thank you for reply!

It seems like there is not problem with my script, but with the scheduler. 
```
File "/opt/netbox/netbox/extras/scripts.py", line 478, in run_script
        script = module.scripts.get(job.name)()
TypeError: 'NoneType' object is not callable 
```
The provided error does not come from my script.

**Reply by candlerb** [reply]
And if you run your script in the way I suggested, does it get silently accepted?

**Reply by drobnymichal** [reply]
No, it is infinitely pending.

And provided error occured again.

**Reply by candlerb** [reply]
Sorry I wasn't clear. I meant: if you temporarily add those extra lines to the top of the file, and you run your script *from the command line*, what happens? (It won't be submitted as a job, and hence it won't be shown as "pending")

**Reply by candlerb** [reply]
Here's the relevant code in /opt/netbox/netbox/extras/scripts.py :

```
    module = ScriptModule.objects.get(pk=job.object_id)
    script = module.scripts.get(job.name)()    # line 478
```

You can test this using [nbshell](https://docs.netbox.dev/en/stable/administration/netbox-shell/).

```
/opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py nbshell
...
>>> module = ScriptModule.objects.get(pk=1)
>>> module
<ScriptModule: add_device_type_components>
>>> module.scripts
{'AddDeviceTypeComponents': <class 'add_device_type_components.AddDeviceTypeComponents'>}
```

So, the first 'get' is a Django ORM get, and it will raise an exception if the given item does not exist (or multiple items match the request).

The second 'get' is a simple dict get. If the key does not exist, then it returns None. And that's why you are getting the error 'NoneType is not callable'.

Personally I would replace line 478 with:

```
script = module.scripts[job.name]()
```

so that you'd get a KeyError instead, which would show the name of the bad item.

From above, you can see that the value it's looking for is the *class* within the script, i.e. the file contains

```
from extras.scripts import Script

class AddDeviceTypeComponents(Script):
    ....
```

So my guess is that the structure of your custom script is wrong. 

Try replacing your script with this one: https://github.com/netbox-community/customizations/blob/master/scripts/add_device_type_components.py
Does it work?

Alternatively, if you already have a job in the database which references a non-existent script, you may need to delete it. Changing line 478 of /opt/netbox/netbox/extras/scripts.py in the way I suggested above may make it clearer what the problem is.

**Comment by kkthxbye-code** [comment]
It's likely that your rqworker can't load the script because it is not present on the machine/container that runs the rqworker. Please verify that the script file is actually present on the machine/container that runs rqworker.
